### PR TITLE
Improved Python docs

### PR DESCRIPTION
**Pull request by tdas**
- Added examples for each method
- Added param and return docs for more complicated methods
- Added version information for all classes and methods

### CODE REVIEW

File: python/delta/tables.py [hunk 1]
```diff
@@ -67,17 +81,26 @@ def delete(self, condition=None):
...
         Update data from the table on the rows that match the given ``condition``,
         which performs the rules defined by ``set``.
 
+        Example::
+
+            # condition using SQL formatted string
+            deltaTable.update("eventType = 'clck'", { "eventType": "'click'" } )
+
+            # condition using Spark SQL functions
+            deltaTable.update(col("eventType") == "clck", { "eventType": lit("click") } )
```
Comment (mukulmurthy) [inline python/delta/tables.py]: Maybe use named parameters for these examples to make it clearer?

File: python/delta/tables.py [hunk 2]
```diff
@@ -25,17 +25,23 @@
 class DeltaTable(object):
     """
         Main class for programmatically interacting with Delta tables.
-        You can create DeltaTable instances using the class methods.
+        You can create DeltaTable instances using the path of the Delta table.::
```
Comment (mukulmurthy) [inline python/delta/tables.py]: What's this ".::" mean?

File: python/delta/tables.py [hunk 3]
```diff
@@ -96,7 +120,54 @@ def merge(self, source, condition):
...
+                {
+                  "date": col("updates.date"),
+                  "eventId": col("updates.eventId"),
+                  "data": col("updates.data"),
+                  "count": lit("1")
+                }
+              ).execute()
+
+        :param source: Condition to match
+        :type source: str or pyspark.sql.Column
```
Comment (mukulmurthy) [inline python/delta/tables.py]: Are the above two lines correct? Isn't it a DataFrame?

File: python/delta/tables.py [hunk 4]
```diff
@@ -110,15 +181,21 @@ def merge(self, source, condition):
...
+        empty DataFrame on successful completion.
+
+        Example::
 
+            deltaTable.vacuum()     # vacuum files not required by versions more than 7 days old
+
+            deltaTable.vacuum(100)  # vacuum files not required by versions more than 100 hours old
+
+        :param retentionHours: Optional number of hours retain history. If not specified, then the
+                               default retention period of 7 days will be used.
```
Comment (mukulmurthy) [inline python/delta/tables.py]: I get why it's this way, but it's still super confusing that the param is in hours and the default is days. Maybe say it in both hours and days? "default retention period of 168 hours (7 days) will be used"?

File: python/delta/tables.py [hunk 5]
```diff
@@ -127,12 +204,21 @@ def vacuum(self, retentionHours=None):
...
-        :param limit:
-        :return: DataFrame of the history
+
+        Example::
+
+            fullHistoryDF = deltaTable.history()    # get the full history of the table
+
+            lastOperationDF = deltaTable.history(1) # get the last operation
+
+        :param limit: Optional, number of latest commits to returns in the history.
```
Comment (mukulmurthy) [inline python/delta/tables.py]: for both history and vacuum, is it common in python to not specify the :type if it's an int?

File: python/delta/tables.py [hunk 6]
```diff
@@ -127,12 +204,21 @@ def vacuum(self, retentionHours=None):
...
+        :return: table's commit history
```
Comment (mukulmurthy) [inline python/delta/tables.py]: Maybe some more information on what to do with this or parse this, or where they can find what the format of this is.

File: python/delta/tables.py [hunk 7]
```diff
@@ -167,10 +268,18 @@ def convertToDelta(cls, sparkSession, identifier, partitionSchema=None):
         return jdt
 
     @classmethod
+    @since(0.4)
     def forPath(cls, sparkSession, path):
         """
         Create a DeltaTable for the data at the given `path` using the given SparkSession.
 
+        :return: DeltaTable representing the converted Delta table
```
Comment (mukulmurthy) [inline python/delta/tables.py]: Nit: docs for path parameter

File: python/delta/tables.py [hunk 8]
```diff
@@ -224,114 +333,206 @@ def _condition_to_jcolumn(cls, condition, argname="'condition'"):
...
         """
         Update a matched table row based on the rules defined by ``set``.
         If a ``condition`` is specified, then it must evaluate to true for the row to be updated.
 
+        See :py:class:`~delta.tables.DeltaMergeBuilder` for complete usage details.
+
+        :param condition: Optional condition of the update
+        :type condition: str or pyspark.sql.Column
+        :param set: Defines the rules of setting the values of columns that need to be updated.
+                    *Note: This param is not Optional.* Default value None is present to allow
```
Comment (mukulmurthy) [inline python/delta/tables.py]: That's... pretty confusing. I'd say required instead of not optional, and I hope we throw a good error for this.

Comment (tdas) [reply python/delta/tables.py]: we do.

File: python/delta/tables.py (hunk 1 above)
Comment (tdas) [reply python/delta/tables.py]: what is unclear about using positional arguments here?

File: python/delta/tables.py (hunk 2 above)
Comment (tdas) [reply python/delta/tables.py]: sphinx syntax to say the next lines is a code snippet and therefore format it.

File: python/delta/tables.py (hunk 3 above)
Comment (tdas) [reply python/delta/tables.py]: my mistake. you are right.

File: python/delta/tables.py (hunk 5 above)
Comment (tdas) [reply python/delta/tables.py]: I explicitly added type for only he cases where its more complicated (e.g. conditions with both str and Column).

File: python/delta/tables.py (hunk 1 above)
Comment (mukulmurthy) [reply python/delta/tables.py]: It's not unclear, but I found the merge example with the named args a little easier to read

File: python/delta/tables.py (hunk 6 above)
Comment (tdas) [reply python/delta/tables.py]: done.

File: python/delta/tables.py (hunk 4 above)
Comment (tdas) [reply python/delta/tables.py]: done.

File: docs/python/conf.py [hunk 9]
```diff
@@ -77,5 +77,5 @@
 # Fix for doc generation to work on older version of sphinx as well.
 master_doc = 'index'
 
-# Display the classes in the generated in the same order as the classes appear in the source files.
+# Display the classes in the generated in the same order as the classes appear in the source files.`
```
Comment (mukulmurthy) [inline docs/python/conf.py]: grammar here. also make sure you meant to put that character at the end

File: python/delta/tables.py [hunk 10]
```diff
@@ -96,7 +124,54 @@ def merge(self, source, condition):
...
+              ).execute()
+
+        :param source: Source DataFrame
+        :type source: pyspark.sql.DataFrame
+        :param condition: Condition to match sources rows with the Delta table rows.
+        :type condition: str or pyspark.sql.Column
+
+        :return: builder object to specify whether to update, delete or insert rows based on
+                 whether the condition matched or not
+        :rtype: :py:class:`delta.tables.DeltaMergeBuilder`
```
Comment (zsxwing) [inline python/delta/tables.py]: nit: missing `~`

**Pull request closed without merging**